beacon-chain/slasher/chunks.go

```diff
@@ -0,0 +1,553 @@
+	// a. If the source epoch is 0, this would correspond to a, which means there is nothing
+	// more to update.
+	if sourceEpoch == 0 {
+		return
+	}
+	var difference types.Epoch
+	if currentEpoch > m.params.historyLength {
+		difference = currentEpoch - m.params.historyLength
+	}
+	if sourceEpoch <= difference {
+		return
+	}
+	epoch = sourceEpoch.Sub(1)
+	exists = true
+	return
+}
+
+// StartEpoch given a source epoch and current epoch, determines the start epoch of
+// a max span chunk for use in chunk updates. The source epoch cannot be >= the current epoch.
+func (m *MaxSpanChunksSlice) StartEpoch(
+	sourceEpoch, currentEpoch types.Epoch,
+) (epoch types.Epoch, exists bool) {
+	if sourceEpoch >= currentEpoch {
+		return
+	}
+	// Given max spans is a list of max targets for source epochs, the precondition is that
+	// every attestation's source epoch must be < than its target epoch. So the start epoch
+	// for updates is given as source epoch + 1.
+	epoch = sourceEpoch.Add(1)
+	exists = true
+	return
+}
+
+// NextChunkStartEpoch given an epoch, determines the start epoch of the next chunk. For min
+// span chunks, this will be the last epoch of chunk index = (current chunk - 1). For example:
+//
+//                       chunk0     chunk1     chunk2
+//                         |          |          |
+//  max_spans_val_i = [[-, -, -], [-, -, -], [-, -, -]]
+//
+// If C = chunkSize is 3 epochs per chunk, and we input start epoch of chunk 1 which is 3. The next start
+// epoch is the last epoch of chunk 0, which is epoch 2. This is computed as:
+//
+//  last_epoch(chunkIndex(startEpoch)-1)
+//  last_epoch(chunkIndex(3) - 1)
+//  last_epoch(1 - 1)
+//  last_epoch(0)
+//  2
+func (m *MinSpanChunksSlice) NextChunkStartEpoch(startEpoch types.Epoch) types.Epoch {
+	prevChunkIdx := m.params.chunkIndex(startEpoch)
+	if prevChunkIdx > 0 {
+		prevChunkIdx--
+	}
+	return m.params.lastEpoch(prevChunkIdx)
+}
+
+// NextChunkStartEpoch given an epoch, determines the start epoch of the next chunk. For max
+// span chunks, this will be the start epoch of chunk index = (current chunk + 1). For example:
+//
+//                       chunk0     chunk1     chunk2
+//                         |          |          |
+//  max_spans_val_i = [[-, -, -], [-, -, -], [-, -, -]]
+//
+// If C = chunkSize is 3 epochs per chunk, and we input start epoch of chunk 1 which is 3. The next start
+// epoch is the start epoch of chunk 2, which is epoch 6. This is computed as:
+//
+//  first_epoch(chunkIndex(startEpoch)+1)
+//  first_epoch(chunkIndex(3)+1)
+//  first_epoch(1 + 1)
+//  first_epoch(2)
+//  6
+func (m *MaxSpanChunksSlice) NextChunkStartEpoch(startEpoch types.Epoch) types.Epoch {
+	return m.params.firstEpoch(m.params.chunkIndex(startEpoch) + 1)
+}
+
+// Given a validator index and epoch, retrieves the target epoch at its specific
+// index for the validator index + epoch pair in a min/max span chunk.
```

Comment:
There doesn't seem to be any epoch *pair* here, at least not in the parameters list. Maybe something changed?